File: Engine/LeviathanMathematics/Vec2.cs
using System.Numerics;

namespace Leviathan.Mathematics
{
	public struct Vec2
	{
		public static Vec2 Zero { get; } = new(0, 0);
		public static Vec2 Half { get; } = new(0.5f, 0.5f);
		public static Vec2 One { get; } = new(1, 1);
		public static Vec2 Right { get; } = new(1, 0);
		public static Vec2 Left { get; } = new(-1, 0);
		public static Vec2 Up { get; } = new(0, 1);
		public static Vec2 Down { get; } = new(0, -1);
		public static Vec2 PositiveInfinity { get; } = new(float.PositiveInfinity, float.PositiveInfinity);
		public static Vec2 NegativeInfinity { get; } = new(float.NegativeInfinity, float.NegativeInfinity);

		public float Magnitude => MathF.Sqrt(SqrMagnitude);
		public float SqrMagnitude => x * x + y * y;
		public Vec2 Normalized
		{
			get
			{
				Vec2 copy = this;
				copy.Normalise();
				
				return copy;
			}
		}

		public float this[int _index] => _index switch
		{
			0 => x,
			1 => y,
			_ => throw new IndexOutOfRangeException("Vec2 only has 2 components")
		};

		public float x = 0.0f;
		public float y = 0.0f;

		public Vec2()
		{
			x = 0;
			y = 0;
		}

		public Vec2(float _x, float _y)
		{
			x = _x;
			y = _y;
		}

		public Vec2(Vector2 _vector2)
		{
			x = _vector2.X;
			y = _vector2.Y;
		}

		public void Normalise()
		{
			float length = Magnitude;
			if(length == 0)
			{
				x = 0;
				y = 0;
			}
			else
			{
				x /= length;
				y /= length;
			}
		}

		public void Set(float _x, float _y)
		{
			x = _x;
			y = _y;
		}

		public bool Equals(Vec2 _other) => x.Equals(_other.x) && y.Equals(_other.y);

		public override bool Equals(object? _obj) => _obj is Vec2 other && Equals(other);

		public override int GetHashCode() => HashCode.Combine(x, y);

		public override string ToString() => $"({x}, {y})";

		public static float Angle(Vec2 _vector) => MathF.Atan2(_vector.y, _vector.x);

		public static Vec2 ClampMagnitude(Vec2 _vector, float _maxMagnitude)
		{
			if(_vector.Magnitude <= _maxMagnitude)
				return _vector;

			_vector.Normalise();

			return _vector * _maxMagnitude;
		}

		//Distance
		public static float Distance(Vec2 _a, Vec2 _b) => (_a - _b).Magnitude;

		public static float Dot(Vec2 _lhs, Vec2 _rhs) => _lhs.x * _rhs.x + _lhs.y * _rhs.y;

		public static Vec2 Lerp(Vec2 _a, Vec2 _b, float _t)
		{
			_t = LMath.Clamp01(_t);

			return _a * (1 - _t) + _b * _t;
		}

		public static Vec2 LerpUnclamped(Vec2 _a, Vec2 _b, float _t) => _a * (1 - _t) + _b * _t;

		public static Vec2 Min(Vec2 _a, Vec2 _b) => new(LMath.Min(_a.x, _b.x), LMath.Min(_a.y, _b.y));
		public static Vec2 Max(Vec2 _a, Vec2 _b) => new(LMath.Max(_a.x, _b.x), LMath.Max(_a.y, _b.y));

		public static Vec2 Reflect(Vec2 _direction, Vec2 _normal)
		{
			float dot = Dot(_direction, _normal);
			
			return _direction - (2 * dot * _normal);
		}

		public static Vec2 Perpendicular(Vec2 _vector) => new(-_vector.y, _vector.x);

		public static Vec2 Rotate(Vec2 _vec, float _angle)
		{
			float angle = _angle * LMath.DEG_2_RAD;

			return new Vec2
			{
				x = _vec.x * MathF.Cos(angle) - _vec.y * MathF.Sin(angle),
				y = _vec.x * MathF.Sin(angle) + _vec.y * MathF.Cos(angle)
			};
		}

		public static Vec2 Scale(Vec2 _vec, float _scalar) => _vec * _scalar;

		public static float SignedAngle(Vec2 _vector) => MathF.Atan2(_vector.y, _vector.x) - 180f;

		/// <summary>
		/// This method allows an implicit conversion from System.Numerics.Vector2 to our Vec2 class.
		/// </summary>
		/// <param name="_vec"></param>
		public static implicit operator Vec2(Vector2 _vec) => new Vec2(_vec.X, _vec.Y);

		/// <summary>
		/// This method allows an implicit conversion from our Vec2 class to System.Numerics.Vector2.
		/// </summary>
		/// <param name="_v"></param>
		public static implicit operator Vector2(Vec2 _v) => new Vector2(_v.x, _v.y);

		/// <summary>
		/// This method allows an implicit conversion from our Vec2 class to System.Numerics.Vector2.
		/// </summary>
		/// <param name="_v"></param>
		public static implicit operator Vec3(Vec2 _v) => new Vec3(_v.x, _v.y, 1);

		/// <summary>
		/// This method allows an implicit conversion from our Vec2 class to System.Numerics.Vector2.
		/// </summary>
		/// <param name="_v"></param>
		public static implicit operator Vec2(Vec3 _v) => new Vec2(_v.x, _v.y);

		//Addition Operator Overload +
		public static Vec2 operator +(Vec2 _lhs, Vec2 _rhs) => new Vec2(_lhs.x + _rhs.x, _lhs.y + _rhs.y);

		//Subtraction Operator Overload - 
		public static Vec2 operator -(Vec2 _lhs, Vec2 _rhs) => new Vec2(_lhs.x - _rhs.x, _lhs.y - _rhs.y);

		public static bool operator ==(Vec2 _lhs, Vec2 _rhs) => _lhs.Equals(_rhs);

		public static bool operator !=(Vec2 _lhs, Vec2 _rhs) => !(_lhs == _rhs);

		public static Vec2 operator *(Vec2 _lhs, float _rhs) => new Vec2(_lhs.x * _rhs, _lhs.y * _rhs);

		public static Vec2 operator *(float _lhs, Vec2 _rhs) => new Vec2(_lhs * _rhs.x, _lhs * _rhs.y);

		public static Vec2 operator /(Vec2 _lhs, float _rhs) => new Vec2(_lhs.x / _rhs, _lhs.y / _rhs);
	}
}

File: Framework/Leviamath/Vec2.cpp
#include <Leviathan/Math/Vec2.h>

Vec2::Vec2() : x(0), y(0)
{
}

Vec2::Vec2(const float _val) : x(_val), y(_val)
{
}

Vec2::Vec2(const float _x, const float _y) : x(_x), y(_y)
{
}

Vec2::Vec2(const initializer_list<float> _values)
    : x(0), y(0)
{
    int i = 0;
    for(const float value : _values)
    {
        values[i] = value;
        i++;
    }
}

void Vec2::Set(const float _x, const float _y)
{
    x = _x;
    y = _y;
}

float Vec2::Magnitude() const
{
    return Leviamath::Sqrt(SqrMagnitude());
}

float Vec2::SqrMagnitude() const
{
    return x * x + y * y;
}

Vec2 Vec2::Normalised() const
{
    const float mag = Magnitude();
    Vec2 normal = Vec2(0);

    if(mag != 0)
    {
        normal.x = x / mag;
        normal.y = y / mag;
    }

    return normal;
}

void Vec2::Normalise()
{
    const float mag = Magnitude();

    if(mag != 0)
    {
        x /= mag;
        y /= mag;
    }
    else
    {
        x = y = 0;
    }
}

float Vec2::Angle(const Vec2 _vec)
{
    return Leviamath::Atan2(_vec.y, _vec.x);
}

Vec2 Vec2::ClampMagnitude(Vec2 _vec, const float _maxMag)
{
    if(_vec.Magnitude() <= _maxMag)
        return _vec;

    _vec.Normalise();
    return _vec * _maxMag;
}

float Vec2::Distance(const Vec2 _lhs, const Vec2 _rhs)
{
    return (_lhs - _rhs).Magnitude();
}

float Vec2::Dot(const Vec2 _lhs, const Vec2 _rhs)
{
    return _lhs.x * _rhs.x + _lhs.y * _rhs.y;
}

Vec2 Vec2::Lerp(const Vec2 _a, const Vec2 _b, float _t)
{
    _t = Leviamath::Clamp01(_t);

    return _a * (1 - _t) + _b * _t;
}

Vec2 Vec2::LerpUnclamped(const Vec2 _a, const Vec2 _b, const float _t)
{
    return _a * (1 - _t) + _b * _t;
}

Vec2 Vec2::Min(const Vec2 _a, const Vec2 _b)
{
    return { Leviamath::Min(_a.x, _b.x), Leviamath::Min(_a.y, _b.y) };
}

Vec2 Vec2::Max(const Vec2 _a, const Vec2 _b)
{
    return { Leviamath::Max(_a.x, _b.x), Leviamath::Max(_a.y, _b.y) };
}

Vec2 Vec2::Normalise(const Vec2 _vec)
{
    const float mag = _vec.Magnitude();
    Vec2 normal = _vec;

    if(mag != 0)
    {
        normal.x = _vec.x / mag;
        normal.y = _vec.y / mag;
    }

    return normal;
}

Vec2 Vec2::Reflect(const Vec2 _incoming, const Vec2 _normal)
{
    const float dot = Dot(_incoming, _normal);

    return _incoming - 2 * dot * _normal;
}

Vec2 Vec2::Perpendicular(const Vec2 _vec)
{
    return { -_vec.y, _vec.x };
}

Vec2 Vec2::Rotate(const Vec2 _vec, const float _angle)
{
    const float angle = _angle * DEG_2_RAD;

    return
    {
        _vec.x * Leviamath::Cos(angle) - _vec.y * Leviamath::Sin(angle),
        _vec.y * Leviamath::Sin(angle) + _vec.y * Leviamath::Cos(angle)
    };
}

Vec2 Vec2::Scale(const Vec2 _vec, const float _scalar)
{
    return _vec * _scalar;
}

Vec2 Vec2::SignedAngle(const Vec2 _vec)
{
    return Leviamath::Atan2(_vec.y, _vec.x) - 180.f;
}

Vec2::operator Vector2() const
{
    return { x, y };
}

Vec2::operator b2Vec2() const
{
    return { x, y };
}

Vec2 operator+(const Vec2 _rhs, const Vec2 _lhs)
{
    return { _rhs.x + _lhs.x, _rhs.y + _lhs.y };
}

Vec2 operator-(const Vec2 _rhs, const Vec2 _lhs)
{
    return { _rhs.x - _lhs.x, _rhs.y - _lhs.y };
}

Vec2 operator*(const Vec2 _rhs, const float _lhs)
{
    return { _rhs.x * _lhs, _rhs.y * _lhs };
}

Vec2 operator*(const float _rhs, const Vec2 _lhs)
{
    return { _rhs * _lhs.x, _rhs * _lhs.y };
}

Vec2 operator/(const Vec2 _rhs, const float _lhs)
{
    return { _rhs.x / _lhs, _rhs.y / _lhs };
}

bool operator==(const Vec2 _rhs, const Vec2 _lhs)
{
    return Leviamath::Approximately(_rhs.x, _lhs.x) && Leviamath::Approximately(_rhs.y, _lhs.y);  // NOLINT(clang-diagnostic-float-conversion)
}

bool operator!=(const Vec2 _rhs, const Vec2 _lhs)
{
    return !(_rhs == _lhs);
}
